src/server/utils/validation/validators.ts

```diff
@@ -0,0 +1,138 @@
+import { definitionConverters, ViewDefinition, ViewDefinitionVersion } from "../../../common/view-definitions";
+import { ViewDefinitionConverter } from "../../../common/view-definitions/view-definition-converter";
+import { checkAccess } from "../datacube-guard/datacube-guard";
+import { SettingsManager } from "../settings-manager/settings-manager";
+
+export class ValidationError extends Error {
+  constructor(message: string, public code = 400) {
+    super(message);
+  }
+}
+
+export function isValidationError(e: Error): e is ValidationError {
+  return e instanceof ValidationError;
+}
+
+function convertToClientAppSettings(appSettings: AppSettings): ClientAppSettings {
+  return deserialize(serialize(appSettings));
+}
+
+function convertToClientDataCube(cube: QueryableDataCube): ClientDataCube {
+  return {
+    ...cube,
+    timeAttribute: cube.timeAttribute && cube.timeAttribute.name
+  };
+}
+
+export function createEssence(
+  viewDefinition: ViewDefinition,
+  converter: ViewDefinitionConverter<ViewDefinition, Essence>,
+  dataCube: QueryableDataCube,
+  appSettings: AppSettings
+): Essence {
+  const clientDataCube = convertToClientDataCube(dataCube);
+  const clientAppSettings = convertToClientAppSettings(appSettings);
+
+  try {
+    return converter.fromViewDefinition(viewDefinition, clientAppSettings, clientDataCube);
+  } catch ({ message }) {
+    throw new ValidationError(`invalid viewDefinition object: ${message}`);
+  }
+}
+
+export function parseExpression(req: Request): Expression {
+  try {
+    return Expression.fromJS(req.body.expression);
+  } catch (e) {
+    throw new ValidationError(`bad expression: ${e.message}`);
+  }
+}
+
+export function parseTimezone(req: Request): Timezone | null {
+  const { timezone } = req.body;
+  if (isNil(timezone)) return null;
+  if (typeof timezone !== "string") {
+    throw new ValidationError("timezone must be a string");
+  }
+  try {
+    return Timezone.fromJS(timezone);
+  } catch (e) {
+    throw new ValidationError(`bad timezone: ${e.message}`);
+  }
+}
+
+export function parseViewDefinitionConverter(req: Request): ViewDefinitionConverter<ViewDefinition, Essence> {
+  const { viewDefinitionVersion } = req.body;
+  if (typeof viewDefinitionVersion !== "string") {
+    throw new ValidationError("must have a viewDefinitionVersion");
+  }
+  const converter = definitionConverters[viewDefinitionVersion as ViewDefinitionVersion];
+
+  if (converter == null) {
+    throw new ValidationError("unsupported viewDefinitionVersion value");
+  }
+
+  return converter;
+}
+
+export function parseViewDefinition(req: Request): ViewDefinition {
+  const { viewDefinition } = req.body;
+  if (typeof viewDefinition !== "object") {
+    throw new ValidationError("viewDefinition must be an object");
+  }
+
+  return viewDefinition;
+}
+
+export async function parseDataCube(req: Request, settings: Pick<SettingsManager, "getSources">): Promise<QueryableDataCube> {
```

Comment:
General idea: I would keep functions closer to data type they operate. Now the validators is a bag for everything.

Reply:
But they operate on the same data type - Request. They produce different types.

If we would transpose this, we would get `parseFromRequest` in `data-cube.ts`, is it better?